lib/cosmos/tools/cmd_tlm_server/replay_backend.rb

```diff
@@ -0,0 +1,372 @@
+            if packet_log_reader == 'DEFAULT'
+              @packet_log_reader = @default_packet_log_reader
+            else
+              packet_log_reader = Cosmos.require_class(packet_log_reader)
+              @packet_log_reader = packet_log_reader_class.new
+            end
+          elsif !packet_log_reader.nil?
+            # Instantiated object
+            @packet_log_reader = packet_log_reader
+          end # Else use existing
+
+          @log_filename = filename
+          @log_directory = File.dirname(@log_filename)
+          @log_directory << '/' unless @log_directory[-1..-1] == '\\'
+
+          System.telemetry.reset
+          @cancel = false
+          @progress = 0
+          @status = "Analyzing: #{@progress}%"
+          start_config_name = System.configuration_name
+          config_change_success, config_error = Cosmos.check_log_configuration(@packet_log_reader, @log_filename)
+          if System.configuration_name != start_config_name
+            @config_change_callback.call() if @config_change_callback
+          end
+          @packet_offsets = @packet_log_reader.packet_offsets(@log_filename, lambda {|percentage| 
+            progress_int = (percentage * 100).to_i
+            if @progress != progress_int
+              @progress = progress_int
+              @status = "Analyzing: #{@progress}%"
+            end          
+            @cancel
+          })
+          @playback_index = 0
+          @playback_max_index = @packet_offsets.length
+          @packet_log_reader.open(@log_filename)
+
+          if @cancel
+            @packet_log_reader.close
+            @log_filename = ''
+            @packet_offsets = []
+            @playback_index = 0
+            @start_time = ''
+            @current_time = ''
+            @end_time = ''
+          else
+            packet = read_at_index(@packet_offsets.length - 1, :FORWARD)
+            @end_time = packet.received_time.formatted(true, 3, true) if packet and packet.received_time
+            packet = read_at_index(0, :FORWARD)
+            @start_time = packet.received_time.formatted(true, 3, true) if packet and packet.received_time            
+          end
+        rescue Exception => error
+          Logger.error "Error in Analysis Thread\n#{error.formatted}"
+        ensure
+          @status = 'Stopped'
+          @playing = false
+          @playback_sleeper = nil
+          @thread = nil
+        end          
+      end 
+    end
+
+    # Get current replay status
+    #
+    # status, playback_delay, filename, file_start, file_current, file_end, file_index, file_max_index
+    def status
+      [@status, 
+        @playback_delay, 
+        @log_filename.to_s, 
+        @start_time, 
+        @current_time, 
+        @end_time, 
+        @playback_index, 
+        @playback_max_index]
+    end
+
+    # Set the replay delay
+    #
+    # @param delay [Float] delay between packets in seconds 0.0 to 1.0, nil = No Delay, -1.0 = REALTIME
```

Comment:
I think we need to document somewhere on the GUI that nil means No Delay and -1 means REALTIME. What's the difference between 0 delay and nil (No Delay)? Could nil be realtime and 0 be no delay? -1 for a value just isn't that intuitive.

Reply:
In the GUI it is a combobox, so doesn't really need to be documented there.   It just needs to be documented with the new API methods.  I agree the values aren't very intuitive and we could make realtime = nil and no delay = 0.  It would just change the code a bit.  I'll look into it.